### PR TITLE
Feature/add youtube api

### PR DESCRIPTION
**Pull request by saihtaungkham**
Adding new Youtube API feature.
- `next_page_token` optional attribute has been added. Use this token for pagination in Youtube.
- Fix some bug in Github Fetching API.
- New attributes have been added in Protocol class in order to support Youtube API.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -38,6 +38,10 @@ Optional Parameters:
                         Paper with code repository password.
   -t GITHUB_ACC_TOKEN, --github_acc_token GITHUB_ACC_TOKEN
                         Github access token.
+  -yk YOUTUBE_DEV_KEY, --youtube_dev_key YOUTUBE_DEV_KEY
+                        Youtube developer key.
+  -np NEXT_PAGE_TOKEN, --next_page_token NEXT_PAGE_TOKEN
```
Comment (rhezaharliman) [inline README.md]: I think it'll be good if we can specify that this only for YouTube? How do you think?
Like `ynpt`? Or is it too long? 🤔

File: README.md [hunk 2]
```diff
@@ -38,6 +38,10 @@ Optional Parameters:
                         Paper with code repository password.
   -t GITHUB_ACC_TOKEN, --github_acc_token GITHUB_ACC_TOKEN
```
Comment (rhezaharliman) [inline README.md]: It's not related to this PR but if we can specify this too, I think it'll be great (Both PWC and Github).

File: src/main/python/mlsearch/api_requester.py [hunk 3]
```diff
@@ -10,18 +11,20 @@
...
         :param  query:      The query for searching.
         :param  init_idx:   The initial pagination index.
         :param  count:      The number of records to be fetched.
+        :param  next_page_token: The current page token for youtube API.
         """
 
         self.params = {'query':query, 'init_idx':init_idx, 
-                            'count':count, 'source': source}
+                            'count':count, 'source': source,
+                            'next_page_token': next_page_token}
```
Comment (rhezaharliman) [inline src/main/python/mlsearch/api_requester.py]: This one too. if we can specify that this token is for YouTube only, it may be great 🤔

File: src/main/python/mlsearch/api_requester.py [hunk 4]
```diff
@@ -70,7 +82,7 @@ def _validate_params(self):
                 raise TypeError(
                     f'Invalid type for {item}. {typ} is expected but ' 
                     f'{type(self.params[item])} is given.')
-
+
```
Comment (rhezaharliman) [inline src/main/python/mlsearch/api_requester.py]: Unnecessary diff?

File: src/main/python/mlsearch/api_requester.py [hunk 5]
```diff
@@ -161,7 +173,58 @@ def _fetch_paperwithcode(self) -> [Protocol]:
...
         self.data['response_code'] = query_result.status_code
-        
+    
+    def _fetch_youtube(self, next_page_token=None) -> [Protocol]:
+        """Fetch the Youtube Repository"""
+        results = []
+        youtube = googleapiclient.discovery.build(
+            self._config.YOUTUBE_SERVICE_NAME, 
+            self._config.YOUTUBE_API_VERSION, 
+            developerKey = self._config.YOUTUBE_DEVELOPER_KEY)
```
Comment (rhezaharliman) [inline src/main/python/mlsearch/api_requester.py:182]: I think there's a chance that the developer key is empty. I think we should handle that kind of error?

File: src/main/scripts/mlsearch [hunk 6]
```diff
@@ -22,6 +22,8 @@ ap.add_argument('-tm', '--timestamp', required=True, help="Timestamp of requesti
 ap.add_argument('-u', '--pwc_user', required=False, help="Paper with code repository user name.")
 ap.add_argument('-p', '--pwc_password', required=False, help="Paper with code repository password.")
 ap.add_argument('-t', '--github_acc_token', required=False, help="Github access token.")
+ap.add_argument('-yk', '--youtube_dev_key', required=False, help="Youtube developer key.")
+ap.add_argument('-np', '--next_page_token', required=False, help="Next page token for Youtube API.")
```
Comment (rhezaharliman) [inline src/main/scripts/mlsearch]: This area too

File: src/main/scripts/mlsearch [hunk 7]
```diff
@@ -35,30 +37,41 @@ def main(event):
         param_names = [
             'query', 'init_idx',
             'count', 'source',
-            'cookies', 'timestamp']
+            'cookies', 'timestamp',
+            'next_page_token']
```
Comment (rhezaharliman) [inline src/main/scripts/mlsearch]: This one too

**Review by rhezaharliman — CHANGES_REQUESTED** [review]
Thanks Sai! I reviewed it. Looks great! Please check my comments. 🙏

File: README.md (hunk 1 above)
Comment (saihtaungkham) [reply README.md]:
```suggestion
  -ynpt NEXT_PAGE_TOKEN, --next_page_token NEXT_PAGE_TOKEN
```
Thanks for the review. I am changing it.

File: README.md (hunk 2 above)
Comment (saihtaungkham) [reply README.md]:
```suggestion
  -gt GITHUB_ACC_TOKEN, --github_acc_token GITHUB_ACC_TOKEN
```

File: src/main/scripts/mlsearch (hunk 6 above)
Comment (saihtaungkham) [reply src/main/scripts/mlsearch]:
```suggestion
ap.add_argument('-u', '--pwc_user', required=False, help="Paper with code repository user name.")
ap.add_argument('-p', '--pwc_password', required=False, help="Paper with code repository password.")
ap.add_argument('-gt', '--github_acc_token', required=False, help="Github access token.")
ap.add_argument('-yk', '--youtube_dev_key', required=False, help="Youtube developer key.")
ap.add_argument('-ynpt', '--next_page_token', required=False, help="Next page token for Youtube API.")
```

File: src/main/python/mlsearch/api_requester.py (hunk 3 above)
Comment (saihtaungkham) [reply src/main/python/mlsearch/api_requester.py]:
```suggestion
                            'y_next_page_token': next_page_token}
```

File: src/main/python/mlsearch/api_requester.py (hunk 4 above)
Comment (saihtaungkham) [reply src/main/python/mlsearch/api_requester.py]:
```suggestion
```

Comment (saihtaungkham) [reply src/main/python/mlsearch/api_requester.py]:
```suggestion

```

File: src/main/scripts/mlsearch (hunk 6 above)
Comment (saihtaungkham) [reply src/main/scripts/mlsearch]:
```suggestion
ap.add_argument('-pu', '--pwc_user', required=False, help="Paper with code repository user name.")
ap.add_argument('-pp', '--pwc_password', required=False, help="Paper with code repository password.")
ap.add_argument('-gt', '--github_acc_token', required=False, help="Github access token.")
ap.add_argument('-yk', '--youtube_dev_key', required=False, help="Youtube developer key.")
ap.add_argument('-yntp', '--next_page_token', required=False, help="Next page token for Youtube API.")
```

File: src/main/scripts/mlsearch (hunk 7 above)
Comment (saihtaungkham) [reply src/main/scripts/mlsearch]:
```suggestion
            'y_next_page_token']
```

File: src/main/python/mlsearch/api_requester.py (hunk 5 above)
Comment (saihtaungkham) [reply src/main/python/mlsearch/api_requester.py:182]: Technically the default `MLT account` will be used instead if the `null` input is assigned from the client. Moreover, if the developer key is empty, invalid or null then the lambda will throw an error back to the client.

File: src/main/scripts/mlsearch [hunk 8]
```diff
@@ -35,30 +37,41 @@ def main(event):
...
         if hp.is_valid_parameters(event, param_names):
             params = hp.parse_parameters(event)
             if params.values():
-                api_request = APIRequest(params['source'], params['query'], params['init_idx'], params['count'])
+                api_request = APIRequest(
+                    params['source'], 
+                    params['query'], 
+                    params['init_idx'], 
+                    params['count'],
+                    params['next_page_token'])
```
Comment (saihtaungkham) [inline src/main/scripts/mlsearch]:
```suggestion
                    params['y_next_page_token'])
```

File: src/main/scripts/mlsearch [hunk 9]
```diff
@@ -35,30 +37,41 @@ def main(event):
...
+                    api_request.youtube_developer_key = event['youtube_developer_key']
                 data = api_request.fetch_data()
                 response_msg = hp.response(
                     message=data.get('content',''),
                     status_code=data.get('response_code'),
                     headers=headers,
-                    optional_attributes={'has_next_page': data.get('has_next_page', False)})
+                    optional_attributes={
+                        'has_next_page': data.get('has_next_page', False),
+                        'next_page_token': data.get('next_page_token', None)})
```
Comment (saihtaungkham) [inline src/main/scripts/mlsearch]:
```suggestion
                        'next_page_token': data.get('y_next_page_token', None)})
```

File: README.md [hunk 10]
```diff
@@ -36,8 +36,12 @@ Optional Parameters:
                         Paper with code repository user name.
   -p PWC_PASSWORD, --pwc_password PWC_PASSWORD
```
Comment (saihtaungkham) [inline README.md]:
```suggestion
  -pp PWC_PASSWORD, --pwc_password PWC_PASSWORD
```

File: README.md [hunk 11]
```diff
@@ -34,10 +34,14 @@ Optional Parameters:
 
   -u PWC_USER, --pwc_user PWC_USER
```
Comment (saihtaungkham) [inline README.md]:
```suggestion
  -pu PWC_USER, --pwc_user PWC_USER
```

File: README.md [hunk 12]
```diff
@@ -51,4 +55,5 @@ api_request = APIRequest(source, query,
     init_idx, count)
 api_request.pwc_auth_info = ('user_name', 'password')
 api_request.github_acc_token = 'token'
-```
\ No newline at end of file
+api_request.youtube_developer_key = 'you_key'
```
Comment (saihtaungkham) [inline README.md]:
```suggestion
api_request.youtube_developer_key = 'your_key'
```

File: src/main/python/mlsearch/api_requester.py [hunk 13]
```diff
@@ -10,18 +11,20 @@
 class APIRequest():
     """For handling the different Valid API requests."""
 
-    def __init__(self, source, query, init_idx, count):
+    def __init__(self, source, query, init_idx, count, next_page_token=None):
```
Comment (saihtaungkham) [inline src/main/python/mlsearch/api_requester.py]:
```suggestion
    def __init__(self, source, query, init_idx, count, y_next_page_token=None):
```

File: src/main/python/mlsearch/api_requester.py [hunk 14]
```diff
@@ -10,18 +11,20 @@
...
-    def __init__(self, source, query, init_idx, count):
+    def __init__(self, source, query, init_idx, count, y_next_page_token=None):
         """
         Initialization for the class.
 
         :param  source:     The API request destination.
         :param  query:      The query for searching.
         :param  init_idx:   The initial pagination index.
         :param  count:      The number of records to be fetched.
+        :param  next_page_token: The current page token for youtube API.
```
Comment (saihtaungkham) [inline src/main/python/mlsearch/api_requester.py]:
```suggestion
        :param  y_next_page_token: The current page token for youtube API.
```

File: src/main/python/mlsearch/api_requester.py [hunk 15]
```diff
@@ -32,7 +35,8 @@ def __init__(self, source, query, init_idx, count):
         self.data = {
             'response_code': 201,
             'content': None,
-            'has_next_page': False}
+            'has_next_page': False,
+            'next_page_token': None}
```
Comment (saihtaungkham) [inline src/main/python/mlsearch/api_requester.py]:
```suggestion
            'y_next_page_token': None}
```

File: src/main/python/mlsearch/api_requester.py [hunk 16]
```diff
@@ -161,7 +172,58 @@ def _fetch_paperwithcode(self) -> [Protocol]:
...
+            self._config.YOUTUBE_SERVICE_NAME, 
+            self._config.YOUTUBE_API_VERSION, 
+            developerKey = self._config.YOUTUBE_DEVELOPER_KEY)
+        request = youtube.search().list(
+            part=self._config.YOUTUBE_PART,
+            maxResults=self.params['count'],
+            order=self._config.YOUTUBE_ORDER,
+            q=self.params['query'],
+            safeSearch=self._config.YOUTUBE_SAFESEARCH,
+            pageToken=next_page_token
```
Comment (saihtaungkham) [inline src/main/python/mlsearch/api_requester.py]:
```suggestion
            pageToken=y_next_page_token
```

File: src/main/python/mlsearch/api_requester.py [hunk 17]
```diff
@@ -10,18 +11,20 @@
...
         :param  query:      The query for searching.
         :param  init_idx:   The initial pagination index.
         :param  count:      The number of records to be fetched.
+        :param  y_next_page_token: The current page token for youtube API.
         """
 
         self.params = {'query':query, 'init_idx':init_idx, 
-                            'count':count, 'source': source}
+                            'count':count, 'source': source,
+                            'y_next_page_token': next_page_token}
```
Comment (saihtaungkham) [inline src/main/python/mlsearch/api_requester.py]:
```suggestion
                            'y_next_page_token': y_next_page_token}
```

File: README.md [hunk 18]
```diff
@@ -32,12 +32,16 @@ Required Parameters:
...
                         Paper with code repository user name.
-  -p PWC_PASSWORD, --pwc_password PWC_PASSWORD
+  -pp PWC_PASSWORD, --pwc_password PWC_PASSWORD
                         Paper with code repository password.
-  -t GITHUB_ACC_TOKEN, --github_acc_token GITHUB_ACC_TOKEN
+  -gt GITHUB_ACC_TOKEN, --github_acc_token GITHUB_ACC_TOKEN
                         Github access token.
+  -yk YOUTUBE_DEV_KEY, --youtube_dev_key YOUTUBE_DEV_KEY
+                        Youtube developer key.
+  -ynpt NEXT_PAGE_TOKEN, --next_page_token NEXT_PAGE_TOKEN
```
Comment (saihtaungkham) [inline README.md]:
```suggestion
  -ynpt NEXT_PAGE_TOKEN, --y_next_page_token NEXT_PAGE_TOKEN
```

File: src/main/python/mlsearch/api_requester.py [hunk 19]
```diff
@@ -161,7 +172,58 @@ def _fetch_paperwithcode(self) -> [Protocol]:
             self.data['content'] = [proto.to_JSON() for proto in results]
 
         self.data['response_code'] = query_result.status_code
-        
+    
+    def _fetch_youtube(self, next_page_token=None) -> [Protocol]:
```
Comment (saihtaungkham) [inline src/main/python/mlsearch/api_requester.py]:
```suggestion
    def _fetch_youtube(self, y_next_page_token=None) -> [Protocol]:
```

File: src/main/python/mlsearch/api_requester.py [hunk 20]
```diff
@@ -161,7 +172,58 @@ def _fetch_paperwithcode(self) -> [Protocol]:
...
+                        ).get('liveBroadcastContent', None),
+                    'published_datetime': item.get(
+                        'snippet',dict({'publishedAt': None})
+                        ).get('publishedAt', None),
+                    'thumbnails': item.get(
+                        'snippet',dict({'thumbnails': None})
+                        ).get('thumbnails', None),
+                }
+                results.append(Protocol(data))
+            self.data['next_page_token'] = response.get('nextPageToken', None)
```
Comment (saihtaungkham) [inline src/main/python/mlsearch/api_requester.py]:
```suggestion
            self.data['y_next_page_token'] = response.get('nextPageToken', None)
```

File: src/main/python/mlsearch/api_requester.py [hunk 21]
```diff
@@ -170,7 +232,10 @@ def fetch_data(self) -> json:
                 self._fetch_paperwithcode()
 
             if self.params.get('source', '') == 'github':
-                responses = self._fetch_github()
+                self._fetch_github()
+
+            if self.params.get('source', '') == 'youtube':
+                self._fetch_youtube(self.params.get('next_page_token', None))
```
Comment (saihtaungkham) [inline src/main/python/mlsearch/api_requester.py]:
```suggestion
                self._fetch_youtube(self.params.get('y_next_page_token', None))
```

File: src/main/python/mlsearch/helper.py [hunk 22]
```diff
@@ -40,7 +40,8 @@ def parse_parameters(event):
     :return:            dict(
                             'query', 'init_idx',
                             'count', 'source',
-                            'cookies', 'timestamp')
+                            'cookies', 'timestamp',
+                            'next_page_token')
```
Comment (saihtaungkham) [inline src/main/python/mlsearch/helper.py]:
```suggestion
                            'y_next_page_token')
```

File: src/main/python/mlsearch/helper.py [hunk 23]
```diff
@@ -50,6 +51,7 @@ def parse_parameters(event):
         param['source'] = event['source']
         param['cookies'] = event['cookies']
         param['timestamp'] = event['timestamp']
+        param['next_page_token'] = event['next_page_token']
```
Comment (saihtaungkham) [inline src/main/python/mlsearch/helper.py]:
```suggestion
        param['y_next_page_token'] = event['y_next_page_token']
```

File: src/main/scripts/mlsearch [hunk 24]
```diff
@@ -19,9 +19,11 @@ ap.add_argument('-c', '--count', required=True, help="Total number of results to
...
 ap.add_argument('-ck', '--cookies', required=True, help="Cookies of current user.")
 ap.add_argument('-tm', '--timestamp', required=True, help="Timestamp of requesting API.")
-ap.add_argument('-u', '--pwc_user', required=False, help="Paper with code repository user name.")
-ap.add_argument('-p', '--pwc_password', required=False, help="Paper with code repository password.")
-ap.add_argument('-t', '--github_acc_token', required=False, help="Github access token.")
+ap.add_argument('-pu', '--pwc_user', required=False, help="Paper with code repository user name.")
+ap.add_argument('-pp', '--pwc_password', required=False, help="Paper with code repository password.")
+ap.add_argument('-gt', '--github_acc_token', required=False, help="Github access token.")
+ap.add_argument('-yk', '--youtube_dev_key', required=False, help="Youtube developer key.")
+ap.add_argument('-yntp', '--next_page_token', required=False, help="Next page token for Youtube API.")
```
Comment (saihtaungkham) [inline src/main/scripts/mlsearch]:
```suggestion
ap.add_argument('-yntp', '--y_next_page_token', required=False, help="Next page token for Youtube API.")
```

File: src/main/scripts/mlsearch [hunk 25]
```diff
@@ -71,7 +84,8 @@ if __name__ == "__main__":
         'count': args['count'],
         'source': args['source'],
         'cookies': args['cookies'],
-        'timestamp': args['timestamp']
+        'timestamp': args['timestamp'],
+        'next_page_token': args['next_page_token']
```
Comment (saihtaungkham) [inline src/main/scripts/mlsearch]:
```suggestion
        'y_next_page_token': args['y_next_page_token']
```

File: src/main/scripts/mlsearch [hunk 26]
```diff
@@ -35,30 +37,41 @@ def main(event):
...
+                    api_request.youtube_developer_key = event['youtube_developer_key']
                 data = api_request.fetch_data()
                 response_msg = hp.response(
                     message=data.get('content',''),
                     status_code=data.get('response_code'),
                     headers=headers,
-                    optional_attributes={'has_next_page': data.get('has_next_page', False)})
+                    optional_attributes={
+                        'has_next_page': data.get('has_next_page', False),
+                        'next_page_token': data.get('y_next_page_token', None)})
```
Comment (saihtaungkham) [inline src/main/scripts/mlsearch]:
```suggestion
                        'y_next_page_token': data.get('y_next_page_token', None)})
```

**Review by saihtaungkham — COMMENTED** [review]
@rhezaharliman I have fixed according to your comments. Please check it out. Sorry for inline commit. I should have fixed it locally. 😰

File: src/main/python/mlsearch/api_requester.py (hunk 5 above)
Comment (rhezaharliman) [reply src/main/python/mlsearch/api_requester.py:182]: Thanks, Sai. I mean, since we already know there's a potential error thrown by lambda, I think it may be hard to debug when it happens because the error comes from lambda. If we take care of that error, we may know the problem faster.
But since
> the default MLT account will be used instead if the null input is assigned from the client

I think it'll be fine.

**Review by rhezaharliman — APPROVED** [review]
Thanks Sai! Looks good to me. ( squash seems necessary 😅 )
